alibi_detect/cd/tensorflow/spot_the_diff.py

```diff
@@ -0,0 +1,203 @@
+            for reservoir sampling {'reservoir_sampling': n} is passed.
+        preprocess_fn
+            Function to preprocess the data before computing the data drift metrics.
+        kernel
+            Differentiable TensorFlow model used to define simmilarity between instances, defaults to Gaussian RBF.
+        n_diffs
+            The number of test locations to use, each corresponding to an interpretable difference.
+        initial_diffs
+            Array used to initialise the diffs that will be learned. Defaults to Gaussian
+            for each feature with equal variance to that of reference data.
+        l1_reg
+            Strength of l1 regularisation to apply to the differences.
+        binarize_preds
+            Whether to test for discrepency on soft (e.g. probs/logits) model predictions directly
+            with a K-S test or binarise to 0-1 prediction errors and apply a binomial test.
+        train_size
+            Optional fraction (float between 0 and 1) of the dataset used to train the classifier.
+            The drift is detected on `1 - train_size`. Cannot be used in combination with `n_folds`.
+        n_folds
+            Optional number of stratified folds used for training. The model preds are then calculated
+            on all the out-of-fold instances. This allows to leverage all the reference and test data
+            for drift detection at the expense of longer computation. If both `train_size` and `n_folds`
+            are specified, `n_folds` is prioritized.
+        retrain_from_scratch
+            Whether the classifier should be retrained from scratch for each set of test data or whether
+            it should instead continue training from where it left off on the previous set.
+        seed
+            Optional random seed for fold selection.
+        optimizer
+            Optimizer used during training of the classifier.
+        learning_rate
+            Learning rate used by optimizer.
+        batch_size
+            Batch size used during training of the classifier.
+        preprocess_batch_fn
+            Optional batch preprocessing function. For example to convert a list of objects to a batch which can be
+            processed by the model.
+        epochs
+            Number of training epochs for the classifier for each (optional) fold.
+        verbose
+            Verbosity level during the training of the classifier. 0 is silent, 1 a progress bar.
+        train_kwargs
+            Optional additional kwargs when fitting the classifier.
+        dataset
+            Dataset object used during training.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+
+        if kernel is None:
+            kernel = GaussianRBF(trainable=True)
+        if initial_diffs is None:
+            initial_diffs = np.random.normal(size=(n_diffs,) + x_ref.shape[1:]) * x_ref.std(0)
```

Comment:
So currently only works for arrays? It's fine to have a separate issue open for extensions to `List[Any]` if it is a non-trivial amount of extra work.

Reply:
Additionally, it is also assuming continuous numerical inputs which of course makes sense as default given the GaussianRBF kernel. It would however be interesting to see how this approach could work with e.g. text data after initially embedding the text if there is evidence for "king - man + woman = queen"-like properties of embedding vectors. This would require an additional preprocessing step before applying the kernel.

Reply:
Well we still support the `preprocess_batch_fn` etc but obvs we need full differentiability from `initial_diffs` to classifier output so they must be in a form that can be handled by torch/tensorflow. So the diffs would indeed be in the processed space rather than raw space that allows any object.

I guess we could allow the user to specify the inverse of `preprocess_batch_fn` to map the diffs back to the raw space. I guess this will often require some custom logic though such as for text we'd need to take an argmax or something to select words.

Reply:
I think we don't really support anything other than arrays since `preprocess_batch_fn` would only be used after we already init the classifier with `x_ref` as an array without preprocessing. As a result adding a `preprocess_batch_fn` would lead to a mismatch in the way `x_ref` is used in `self.mean` and how the instances `x` are used in the forward pass of the classifier since they would have been preprocessed.

Reply:
So if there is a `preprocess_batch_fn` it should already be applied before init of `InterpretableClf` to `x_ref` passed to the classifier?